### ISSUE TITLE
Filter by Tags

### ISSUE THREAD

**Issue by nathanielrindlaub** [tnc-ca-geo/animl-frontend#165]
Once we implement [image-level Tags](https://github.com/tnc-ca-geo/animl-frontend/issues/138#issuecomment-1793160414), we'll need a way for users to filter images by them. I imagine we could create a new section on the filter's panel for filtering by tag, and for each available tag create filter checkboxes for both a positive ("<tag>") and negative ("not-<tag>"), the same way we currently have two checkboxes for "reviewed" and "not-reviewed" (see #90).

**Comment by nathanielrindlaub** [comment]
Also realizing we might want work on that `at least one of the following` language. If two or more checkboxes are checked on, does that mean show/hide images where at least one of the selected tags is present? Or show/hide images where all of the selected tags are present? I had meant the latter but it's not clear from that phrase.

`exact match` I had envisioned meaning show/hide images with this exact set of tags, nothing more nothing less.

**Comment by jue-henry** [comment]
Nattys approach gives the user the option to filter-for or filter-tags, but we run into issues when wish to filter for a certain tag, while filtering out certain tags, which I think would be the primary use case. So we can instead build out a more fully-fledged query builder, where we give the user both options to filter-for and filter-out certain tags. This approach would present to the user 2 dropdowns. The first dropdown will include a list of statements like `must include`, `must exclude`, etc, and the second dropdown will be a list of tags. From there, we can add a new row to this list anytime a user fills out a row, and essentially give them a way to build their own unique query when needed. This approach may be a bit unintuitive initially, but gives the user most flexibility. Will have to consider if this is too technically involved for our user base.

**Comment by nathanielrindlaub** [comment]
@jue-henry can you elaborate on the use case you're thinking of in which a user would want to filter for a certain tag while simultaneously filtering out another?

**Comment by nathanielrindlaub** [comment]
### Updated UX thoughts (again)

Well, it's been a journey. Before implementing one of these advanced search / query builder concepts I took another hard look at the genesis of this ticket, and I think I had misunderstood the reason we needed a "not-reviewed" checkbox option in the first place, and because this ticket is an extension of that misunderstanding I now don't think any of it is relevant or necessary. TL;DR, I think we can replicate the exact UX and logic of Labels filtering and it will meet users' needs just fine.

Here's where I went wrong:

Initially I had assumed that because images can either have a particular tag or not have it, much being "reviewed" or not, in addition to providing a checkbox for users to use to include or not include the tag, we'd also need to provide some way to filter images that _don't_ have that tag (the "negatives" referred to in the first comment on this issue). However, the fact that being "reviewed" is a boolean, true/false, presence/absence state isn't why it needed a "not-reviewed" (negative) checkbox option, it's because in that filter category, there was only one checkbox option initially (the "reviewed" option). 

To understand why this was an issue, remember that when all checkboxes in any particular filter category (e.g. the Deployments filter section or the Labels filter section) are checked, we do nothing. That is we don't pass the DB and query at all - we call `Images.find({})` - and it returns all images. As soon as any checkbox is _unchecked_, however, we send a query that selects for all of the options that remained checked, e.g. `Images.find({ labels: { $in: ['all', 'remaining', 'checked', 'labels']})`. This means that if there were images that didn't have any Labels, those would get returned when all checkboxes are checked because no filtering is happening at all, but as soon as any option is unchecked, those images would not be returned because they aren't being selected for explicitly in the filters (note: technically we do generate a "none" Label filter option but set that aside for a second). This also means that selecting just one option in a filter category (e.g. by clicking "only" zebra), you are able to only see images that have zebra labels and nothing else `Images.find({ labels: { $in: ['zebra']})`

That's all fine and feels intuitive from a user perspective when there are multiple options in a label category, but when there's only one option in a filter category like the "reviewed" category - there's no way for the user to indicate (or for the app to know) whether _all checkboxes in this category are selected, so don't filter at all_ or _only 'reviewed' is selected, so return only images that are 'reviewed'._ That is why we needed that additional "not-reviewed" checkbox: having a second option allowed the "only reviewed" selection to only return reviewed images instead of also representing "include reviewed, but show me everything else too."

Another, related concern is whether or not the filter options in any given filter category are comprehensive or not. By that I mean, is every image in the DB represented by at least one filter option within that category, such that if you were to un-select all options, you'd get back zero images. That is the case for Deployments because every image has to be associated with a deployment, but it's not the case for Labels, because images don't necessarily have to have any labels. If we didn't shim the Labels options with that faux "none" option ("none" is not a real Label), as it stands the filtering logic would _include_ images that don't have any labels only when all checkboxes are selected, but as soon as one option is unselected, those images w/o labels would not get returned. That behavior would be a bit odd and unexpected, so we added the "none" in the Labels filter category, and in the case of the "non-reviewed" option, that's doing double duty as both adding another option to the Reviewed filter category (making selecting only "reviewed" possible, as described above) and serving to represent all of the images that lack "reviewed" = true state, thus making the category comprehensive.

### conclusion

Implement Tag filtering the same way we've implemented Deployment and Label filtering, and be sure to include a "no-tag" option to ensure that the tag category is comprehensive.

**Comment by jue-henry** [comment]
> [@jue-henry](https://github.com/jue-henry) can you elaborate on the use case you're thinking of in which a user would want to filter for a certain tag while simultaneously filtering out another?

I guess what I am thinking isn't exclusive to a combination of filtering for a tag and filtering out a tag simultaneously. This is more an observation based on how we filter for labels, but we don't have any method to filter out for anything. For example, what if we wanted to filter out all `seen` images? Based on our current label system, we would uncheck it but if the image has any other tags, the image would still be included.

**Comment by nathanielrindlaub** [comment]
@jue-henry yeah, that's correct. If you unchecked a `seen` tag and it also had another tag that was still checked on, it would show up. I think for now I'll just implement it that way and perhaps down the road we could implement more advanced query building functionality for both Labels and Tags.

**Comment by nathanielrindlaub** [comment]
- [ ] check that tags can be added to Views, and that when tags are deleted from a Project they also get removed from any views they were a part of.
